pkg/roles/repository.go

```
@@ -0,0 +1,170 @@
+// Copyright 2025 Canonical Ltd.
+// SPDX-License-Identifier: AGPL-3.0
+
+package roles
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+
+	sq "github.com/Masterminds/squirrel"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/monitoring"
+	"github.com/canonical/identity-platform-admin-ui/internal/tracing"
+	"github.com/canonical/identity-platform-admin-ui/pkg/storage"
+)
+
+type RoleRepository struct {
+	db storage.DBClientInterface
+
+	logger  logging.LoggerInterface
+	tracer  tracing.TracingInterface
+	monitor monitoring.MonitorInterface
+}
+
+func (r *RoleRepository) FindRole(ctx context.Context, userID, ID string) (*Role, error) {
+	ctx, span := r.tracer.Start(ctx, "roles.RoleRepository.GetRole")
+	defer span.End()
+
+	row := r.db.Statement().Select("id", "name").
+		From(storage.ROLE_TABLE).
+		Where(sq.Eq{"owner": userID}).
+		QueryRowContext(ctx)
+
+	id, name := "", ""
+
+	err := row.Scan(&id, &name)
+	if err != nil {
+		return nil, fmt.Errorf("unable to scan FindRole result, %w", err)
+	}
+
+	if id == "" {
+		return nil, storage.ErrNotFound
+	}
+
+	return &Role{
+		ID:   id,
+		Name: name,
+	}, nil
+}
+
+func (r *RoleRepository) ListRoles(ctx context.Context, userID string) ([]string, error) {
+	ctx, span := r.tracer.Start(ctx, "roles.RoleRepository.ListRoles")
+	defer span.End()
+
+	rows, err := r.db.Statement().Select("name").
+		From(storage.ROLE_TABLE).
+		Where(sq.Eq{"owner": userID}).
+		Query()
+
+	if errors.Is(err, sql.ErrNoRows) {
```

(Disregard it if I'm wrong):

It seems `Query` does not return such error `sql.ErrNoRows`.

---

I was not able to confirm this. Techinally the Query method is part of the Squirrel library, which in turn invokes methods of the underlying database/sql interface, which should throw that error.
I am currently testing these behavior in particular and should be able to provide more info a little later. I will avoid putting the PR in draft for this and hopefully have something more clear by the end of today

---

Finally I was able to confirm: you are 100% correct, thank you for pointing me in the right direction
Here's the [explanation](https://pkg.go.dev/database/sql#ErrNoRows:~:text=var%20ErrNoRows%20%3D,until%20a%20Scan.)